tool-openssl/verify.cc

```diff
@@ -0,0 +1,177 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/base.h>
+#include <openssl/x509.h>
+#include <openssl/pem.h>
+#include "internal.h"
+
+static const argument_t kArguments[] = {
+        { "-help", kBooleanArgument, "Display option summary" },
+        { "-CAfile", kOptionalArgument, "A file of trusted certificates. The "
+                "file should contain one or more certificates in PEM format." },
+        { "", kOptionalArgument, "" }
+};
+
+// setup_verification_store sets up an X509 certificate store for verification.
+// It configures the store with file and directory lookups. It loads the
+// specified CA file if provided and otherwise uses default locations.
+static X509_STORE *setup_verification_store(std::string CAfile) {
+  bssl::UniquePtr<X509_STORE> store(X509_STORE_new());
+  X509_LOOKUP *lookup;
+
+  if (!store) {
+    return nullptr;
+  }
+
+  if (!CAfile.empty()) {
+    lookup = X509_STORE_add_lookup(store.get(), X509_LOOKUP_file());
+    if (!lookup || !X509_LOOKUP_load_file(lookup, CAfile.c_str(), X509_FILETYPE_PEM)) {
+      fprintf(stderr, "Error loading file %s\n", CAfile.c_str());
+      return nullptr;
+    }
+  }
+
+  // Add default dir path
+  lookup = X509_STORE_add_lookup(store.get(), X509_LOOKUP_hash_dir());
+  if (!lookup || !X509_LOOKUP_add_dir(lookup, NULL, X509_FILETYPE_DEFAULT)) {
+    return nullptr;
+  }
+
+  return store.release();
+}
+
+static int cb(int ok, X509_STORE_CTX *ctx) {
+  int cert_error = X509_STORE_CTX_get_error(ctx);
+  X509 *current_cert = X509_STORE_CTX_get_current_cert(ctx);
+
+  if (!ok) {
+    if (current_cert != NULL) {
+      X509_NAME_print_ex_fp(stderr,
+                         X509_get_subject_name(current_cert),
+                         0, XN_FLAG_ONELINE);
+      fprintf(stderr, "\n");
+    }
+    fprintf(stderr, "%serror %d at %d depth lookup: %s\n",
+               X509_STORE_CTX_get0_parent_ctx(ctx) ? "[CRL path] " : "",
+               cert_error,
+               X509_STORE_CTX_get_error_depth(ctx),
+               X509_verify_cert_error_string(cert_error));
+
+    /*
+     * Pretend that some errors are ok, so they don't stop further
+     * processing of the certificate chain.  Setting ok = 1 does this.
+     * After X509_verify_cert() is done, we verify that there were
+     * no actual errors, even if the returned value was positive.
+     */
+    switch (cert_error) {
+      case X509_V_ERR_NO_EXPLICIT_POLICY:
+        /* fall thru */
+      case X509_V_ERR_CERT_HAS_EXPIRED:
+        /* Continue even if the leaf is a self-signed cert */
+      case X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT:
+        /* Continue after extension errors too */
+      case X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY:
```

Comment:
Is this intentionally different from OpenSSL? There tool does not special case `X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY`

Reply:
Ah good catch, I had added it in for figuring out why our output was different from OpenSSL in some cases. This was the error message that was problematic but this change did not do anything - I just forgot to remove it.